### ISSUE TITLE
Machine Learning: Proper Data Splits vs. Data Leakage

### ISSUE THREAD

**Issue by abubokkor1006** [abubokkor1006/STATLAB#8]
### Discussed in https://github.com/abubokkor1006/STATLAB/discussions/7

<div type='discussions-op-text'>

<sup>Originally posted by **abubokkor1006** March 21, 2025</sup>
# Load necessary libraries
library(caret)
library(e1071)  # For SVM model

# Set seed for reproducibility
set.seed(123)

# Simulate some data
n <- 1000  # Number of data points
p <- 20    # Number of features
data <- as.data.frame(matrix(rnorm(n * p), nrow = n, ncol = p))
target <- as.factor(sample(c(0, 1), n, replace = TRUE))  # Binary target variable

# Combine the features and target into one data frame
dataset <- cbind(data, target)
colnames(dataset)[p+1] <- "target"

# Split data into training (80%) and testing (20%) using caret's createDataPartition
trainIndex <- createDataPartition(dataset$target, p = 0.8, list = FALSE)
trainData <- dataset[trainIndex, ]
testData <- dataset[-trainIndex, ]

# ----------- Train and evaluate a model using simple train-test split -------------

# Train an SVM model on the training set
svm_model <- svm(target ~ ., data = trainData)

# Predict on the test set
svm_predictions <- predict(svm_model, testData)

# Calculate accuracy of the model
simple_accuracy <- mean(svm_predictions == testData$target)
cat("Simple Train-Test Split Accuracy: ", simple_accuracy, "\n")

# ----------- Cross-validation with k-Fold ---------------------------------------

# Perform 10-fold cross-validation using caret's train function
cv_model <- train(target ~ ., data = dataset, method = "svmRadial", trControl = trainControl(method = "cv", number = 10))

# Display cross-validation results
cat("Cross-Validation Accuracy: ", cv_model$results$Accuracy[1], "\n")

# ----------- What happens with improper splitting (Data Leakage) ------------------

# Simulate data leakage by adding a feature that is directly related to the target variable
dataset$leaked_feature <- dataset$target

# Split the dataset again, now with the leaked feature
trainIndex_leak <- createDataPartition(dataset$target, p = 0.8, list = FALSE)
trainData_leak <- dataset[trainIndex_leak, ]
testData_leak <- dataset[-trainIndex_leak, ]

# Train an SVM model with data leakage
svm_model_leak <- svm(target ~ ., data = trainData_leak)

# Predict on the test set with data leakage
svm_predictions_leak <- predict(svm_model_leak, testData_leak)

# Calculate accuracy of the model with data leakage
leak_accuracy <- mean(svm_predictions_leak == testData_leak$target)
cat("Accuracy with Data Leakage (Improper Split): ", leak_accuracy, "\n")

# ----------- Time Series Example (Simulated) -----------------------------------

# Simulate time series data (temporal order)
time_data <- data.frame(
  time = 1:n,
  feature1 = rnorm(n),
  feature2 = rnorm(n),
  target = as.factor(sample(c(0, 1), n, replace = TRUE))
)

# Split into train and test while respecting temporal order
train_data_ts <- time_data[1:800, ]
test_data_ts <- time_data[801:n, ]

# Train an SVM model on time series data
svm_model_ts <- svm(target ~ feature1 + feature2, data = train_data_ts)

# Predict on the test set
svm_predictions_ts <- predict(svm_model_ts, test_data_ts)

# Calculate accuracy on time series data
ts_accuracy <- mean(svm_predictions_ts == test_data_ts$target)
cat("Time Series Test Accuracy (Proper Split): ", ts_accuracy, "\n")

# ----------- Randomized Train-Test Split (Incorrect for Time Series) -------------

# Shuffle data to break temporal order (incorrect for time series data)
shuffled_data <- time_data[sample(nrow(time_data)), ]

# Split shuffled data
train_data_shuffled <- shuffled_data[1:800, ]
test_data_shuffled <- shuffled_data[801:n, ]

# Train an SVM model on shuffled data
svm_model_shuffled <- svm(target ~ feature1 + feature2, data = train_data_shuffled)

# Predict on the shuffled test set
svm_predictions_shuffled <- predict(svm_model_shuffled, test_data_shuffled)

# Calculate accuracy on shuffled data (incorrect split for time series)
shuffled_accuracy <- mean(svm_predictions_shuffled == test_data_shuffled$target)
cat("Shuffled Data Test Accuracy (Improper Split for Time Series): ", shuffled_accuracy, "\n")

# Load necessary libraries for visualization
library(ggplot2)

# Create a data frame to hold the accuracy results
accuracy_data <- data.frame(
  Method = c("Simple Train-Test", "Cross-Validation", "Data Leakage", "Time Series Split", "Shuffled Split"),
  Accuracy = c(simple_accuracy, 
               cv_model$results$Accuracy[1], 
               leak_accuracy, 
               ts_accuracy, 
               shuffled_accuracy)
)

# Create a bar plot to visualize the accuracy results
ggplot(accuracy_data, aes(x = Method, y = Accuracy, fill = Method)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  labs(title = "Model Accuracy Comparison", y = "Accuracy", x = "Method") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  geom_text(aes(label = round(Accuracy, 2)), vjust = -0.3)

# Load necessary libraries
library(caret)
library(e1071)  # For SVM model

# Set seed for reproducibility
set.seed(123)

# Simulate some data
n <- 1000  # Number of data points
p <- 20    # Number of features
data <- as.data.frame(matrix(rnorm(n * p), nrow = n, ncol = p))
target <- as.factor(sample(c(0, 1), n, replace = TRUE))  # Binary target variable

# Combine the features and target into one data frame
dataset <- cbind(data, target)
colnames(dataset)[p+1] <- "target"

# ----------- Case 1: Proper Split (Training and Testing) -------------

# Proper train-test split using caret's createDataPartition
trainIndex <- createDataPartition(dataset$target, p = 0.8, list = FALSE)
trainData <- dataset[trainIndex, ]
testData <- dataset[-trainIndex, ]

# Train an SVM model on the training set
svm_model <- svm(target ~ ., data = trainData)

# Predict on the test set
svm_predictions <- predict(svm_model, testData)

# Calculate accuracy of the model with proper split
proper_accuracy <- mean(svm_predictions == testData$target)
cat("Proper Train-Test Split Accuracy: ", proper_accuracy, "\n")

# ----------- Case 2: Overfitting due to Data Leakage ----------------

# Simulate data leakage by adding a feature that is directly related to the target variable
dataset$leaked_feature <- dataset$target

# Split the dataset again, now with the leaked feature
trainIndex_leak <- createDataPartition(dataset$target, p = 0.8, list = FALSE)
trainData_leak <- dataset[trainIndex_leak, ]
testData_leak <- dataset[-trainIndex_leak, ]

# Train an SVM model with data leakage
svm_model_leak <- svm(target ~ ., data = trainData_leak)

# Predict on the test set with data leakage
svm_predictions_leak <- predict(svm_model_leak, testData_leak)

# Calculate accuracy of the model with data leakage
leak_accuracy <- mean(svm_predictions_leak == testData_leak$target)
cat("Accuracy with Data Leakage (Overfitting): ", leak_accuracy, "\n")
cat("*Warning*: This model is overfitted due to the leaked feature. Performance is likely to be misleading!\n")

# ----------- Case 3: Incorrect Split (Randomized) --------------------

# Randomize data split - This is a mistake when working with time series data!
shuffled_data <- dataset[sample(nrow(dataset)), ]

# Split shuffled data
trainData_shuffled <- shuffled_data[1:800, ]
testData_shuffled <- shuffled_data[801:n, ]

# Train an SVM model on shuffled data
svm_model_shuffled <- svm(target ~ ., data = trainData_shuffled)

# Predict on the shuffled test set
svm_predictions_shuffled <- predict(svm_model_shuffled, testData_shuffled)

# Calculate accuracy on shuffled data
shuffled_accuracy <- mean(svm_predictions_shuffled == testData_shuffled$target)
cat("Accuracy with Randomized Split (Improper for Time Series): ", shuffled_accuracy, "\n")
cat("*Warning*: Shuffling breaks the temporal order! The model's performance is unrealistic for time-series data.\n")

# ----------- Case 4: Temporal Split (Correct for Time Series) ------------

# Simulate time series data (temporal order)
time_data <- data.frame(
  time = 1:n,
  feature1 = rnorm(n),
  feature2 = rnorm(n),
  target = as.factor(sample(c(0, 1), n, replace = TRUE))
)

# Split into train and test while respecting temporal order
train_data_ts <- time_data[1:800, ]
test_data_ts <- time_data[801:n, ]

# Train an SVM model on time series data
svm_model_ts <- svm(target ~ feature1 + feature2, data = train_data_ts)

# Predict on the test set
svm_predictions_ts <- predict(svm_model_ts, test_data_ts)

# Calculate accuracy on time series data
ts_accuracy <- mean(svm_predictions_ts == test_data_ts$target)
cat("Time Series Test Accuracy (Proper Split): ", ts_accuracy, "\n")

# ----------- Case 5: Data Leakage in Time Series (Breaking Temporal Order) ------------

# Shuffle the time series data (incorrect for time series)
shuffled_time_data <- time_data[sample(nrow(time_data)), ]

# Split shuffled time series data
train_data_ts_shuffled <- shuffled_time_data[1:800, ]
test_data_ts_shuffled <- shuffled_time_data[801:n, ]

# Train an SVM model on shuffled time series data
svm_model_ts_shuffled <- svm(target ~ feature1 + feature2, data = train_data_ts_shuffled)

# Predict on the shuffled test set
svm_predictions_ts_shuffled <- predict(svm_model_ts_shuffled, test_data_ts_shuffled)

# Calculate accuracy on shuffled time series data (Incorrect!)
shuffled_ts_accuracy <- mean(svm_predictions_ts_shuffled == test_data_ts_shuffled$target)
cat("Accuracy with Shuffled Time Series (Improper Temporal Split): ", shuffled_ts_accuracy, "\n")
cat("*Warning*: Shuffling time series data ruins the temporal structure and causes unrealistic predictions.\n")

# Load necessary libraries for visualization
library(ggplot2)

# Create a data frame to hold the accuracy results
accuracy_data <- data.frame(
  Method = c("Proper Train-Test Split", 
             "Data Leakage (Overfitting)", 
             "Randomized Split (Improper for Time Series)", 
             "Time Series Split (Proper)", 
             "Shuffled Time Series (Improper)"),
  Accuracy = c(proper_accuracy, 
               leak_accuracy, 
               shuffled_accuracy, 
               ts_accuracy, 
               shuffled_ts_accuracy)
)

# Create a bar plot to visualize the accuracy results
ggplot(accuracy_data, aes(x = Method, y = Accuracy, fill = Method)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  labs(title = "Model Accuracy Comparison Across Different Splits", y = "Accuracy", x = "Method") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  geom_text(aes(label = round(Accuracy, 2)), vjust = -0.3)

# Train SVM model on the original dataset (no leakage)
svm_model <- svm(target ~ ., data = trainData)
svm_predictions_train <- predict(svm_model, trainData)
svm_predictions_test <- predict(svm_model, testData)

# Calculate training and testing accuracy without data leakage
train_accuracy <- mean(svm_predictions_train == trainData$target)
test_accuracy <- mean(svm_predictions_test == testData$target)

# Simulate data leakage by adding a feature that is directly related to the target variable
dataset$leaked_feature <- dataset$target

# Split the dataset again, now with the leaked feature
trainData_leak <- dataset[trainIndex, ]
testData_leak <- dataset[-trainIndex, ]

# Train an SVM model with data leakage
svm_model_leak <- svm(target ~ ., data = trainData_leak)
svm_predictions_train_leak <- predict(svm_model_leak, trainData_leak)
svm_predictions_test_leak <- predict(svm_model_leak, testData_leak)

# Calculate training and testing accuracy with data leakage
train_accuracy_leak <- mean(svm_predictions_train_leak == trainData_leak$target)
test_accuracy_leak <- mean(svm_predictions_test_leak == testData_leak$target)

# Create a data frame to compare training and test accuracy with and without data leakage
accuracy_comparison <- data.frame(
  Method = c("Train Without Leakage", "Test Without Leakage", "Train With Leakage", "Test With Leakage"),
  Accuracy = c(train_accuracy, test_accuracy, train_accuracy_leak, test_accuracy_leak),
  Data = c("Training", "Testing", "Training", "Testing")
)

# Create a bar plot to visualize overfitting
library(ggplot2)
ggplot(accuracy_comparison, aes(x = Method, y = Accuracy, fill = Data)) +
  geom_bar(stat = "identity", position = "dodge") +
  theme_minimal() +
  labs(title = "Overfitting Visualization with and without Data Leakage", 
       y = "Accuracy", 
       x = "Dataset (Training vs Test)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  geom_text(aes(label = round(Accuracy, 2)), vjust = -0.3)

![Image](https://github.com/user-attachments/assets/f2bfcc38-625d-4fb5-af54-c6d50791c464)

![Image](https://github.com/user-attachments/assets/0e8b287e-bb94-4c39-aa81-0a0074fe550f)

![Image](https://github.com/user-attachments/assets/7e09d1db-686b-4429-9e3e-ba3a8d0953a8)</div>

**Comment by abubokkor1006** [comment]
🚀 Machine Learning: Proper Data Splits vs. Data Leakage 🚀

In the world of machine learning, one of the key challenges we face is ensuring that our models are trained and tested correctly to avoid overfitting and biased results. Here’s a quick breakdown of common pitfalls and the importance of proper data handling:

🔑 Proper Train-Test Split: A simple train-test split, where we use one portion of data for training and another for testing, is essential for evaluating model performance. Using this method, I ran a Support Vector Machine (SVM) model on simulated data and found an accuracy of ~53%.

⚖️ Cross-Validation: Instead of a single train-test split, cross-validation helps in better assessing model performance by training and testing on different subsets of data. Using a 10-fold cross-validation, the accuracy dropped slightly to ~50% – but this method provides a more generalized estimate.

⚠️ Data Leakage: Adding features directly related to the target variable (data leakage) can mislead our model into giving perfect results during testing – a major red flag! In my test, a model trained with leaked features achieved 100% accuracy, but that’s not realistic. Warning: This overfitting makes the model unreliable in real-world scenarios.

📊 Time Series Splits: For time-series data, it's essential to maintain the chronological order of data. A proper temporal split gives us a ~57% accuracy. However, shuffling time-series data can break the temporal dependency, resulting in incorrect predictions (accuracy of ~51% in my test).

Takeaways for building robust models:

Ensure data splits are done correctly to avoid overfitting and underperformance.
Cross-validation offers a more reliable estimate.
Be cautious of data leakage – it can make your model appear deceptively accurate.
For time-series, don’t shuffle! Keep the chronological order intact.
Here’s a visual summary of the accuracy results across different methods:

📊 (Include the bar plot or image)

#MachineLearning #DataScience #AI #ModelBuilding #DataLeakage #CrossValidation #TimeSeriesAnalysis #SVM
